src/main/java/net/fabricmc/loom/MixinAnnotationProcessorExtension.java

```diff
@@ -0,0 +1,236 @@
+		}
+
+		extra.set(MIXIN_INFORMATION_CONTAINER, container);
+	}
+
+	private PatternSet add0(SourceSet sourceSet, String refmapName) {
+		PatternSet pattern = new PatternSet().setIncludes(Collections.singletonList("*.mixins.json"));
+		setMixinInformationContainer(sourceSet, new MixinInformationContainer(sourceSet, refmapName, pattern));
+
+		isDefault = false;
+
+		return pattern;
+	}
+
+	/**
+	 * Apply Mixin AP to sourceSet.
+	 * @param sourceSet the sourceSet that applies Mixin AP
+	 * @param refmapName the output ref-map name
+	 */
+	public void add(SourceSet sourceSet, String refmapName, Action<PatternSet> action) {
+		PatternSet pattern = add0(sourceSet, refmapName);
+		action.execute(pattern);
+	}
+
+	public void add(SourceSet sourceSet, String refmapName) {
+		add(sourceSet, refmapName, x -> { });
+	}
+
+	public void add(String sourceSetName, String refmapName, Action<PatternSet> action) {
+		// try to find sourceSet with name sourceSetName in this project
+		SourceSet sourceSet = project.getConvention().getPlugin(JavaPluginConvention.class)
+				.getSourceSets().findByName(sourceSetName);
+
+		if (sourceSet == null) {
+			throw new InvalidUserDataException("No sourceSet " + sourceSetName + " was found");
+		}
+
+		PatternSet pattern = add0(sourceSet, refmapName);
+		action.execute(pattern);
+	}
+
+	public void add(String sourceSetName, String refmapName) {
+		add(sourceSetName, refmapName, x -> { });
+	}
+
+	/**
+	 * Apply Mixin AP to sourceSet with output ref-map name equal to {@code loom.refmapName}.
+	 * @param sourceSet the sourceSet that applies Mixin AP
+	 */
+	public void add(SourceSet sourceSet, Action<PatternSet> action) {
+		LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
+		add(sourceSet, extension.getRefmapName(), action);
+	}
+
+	public void add(SourceSet sourceSet) {
+		add(sourceSet, x -> { });
+	}
+
+	public void add(String sourceSetName, Action<PatternSet> action) {
+		LoomGradleExtension extension = project.getExtensions().getByType(LoomGradleExtension.class);
+		add(sourceSetName, extension.getRefmapName(), action);
+	}
+
+	public void add(String sourceSetName) {
+		add(sourceSetName, x -> { });
+	}
+
+	@NotNull
+	public Stream<SourceSet> getSourceSets(Project project0) {
+		return project0.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().stream()
+				.filter(sourceSet -> {
+					MixinInformationContainer container = getMixinInformationContainer(sourceSet);
+
+					if (container != null) {
+						PatternSet pattern = container.mixinJsonPattern;
+						Stream<String> mixinJsonNames = sourceSet.getResources()
+								.matching(pattern).getFiles().stream().map(File::getName);
+						container.setMixinJsonNames(mixinJsonNames);
+						return true;
+					}
+
+					return false;
+				});
+	}
+
+	@NotNull
+	public Stream<Configuration> getApConfigurations(Project project0, Function<String, String> getApConfigNameFunc) {
```

Comment:
Why pass in project0 when you already have project? I guess this is left over from the days of this being across project.

Reply:
fixed